### PR TITLE
WS URL can be optional when LogBroadcaster is disabled

### PR DESCRIPTION
**Pull request by huangzhen1997**
### Description
WS URL can be optional (empty string) when LogBroadcaster is disabled. If WS URL was not provided, SubscribeFilterLogs should fail with an explicit error

### Tickets:
[BCFR-451](https://smartcontract-it.atlassian.net/browse/BCFR-451)

[BCFR-451]: https://smartcontract-it.atlassian.net/browse/BCFR-451?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: core/chains/evm/config/toml/config.go [hunk 1]
```diff
@@ -964,19 +964,8 @@ func (n *Node) ValidateConfig() (err error) {
...
-	if n.WSURL == nil {
-		if !sendOnly {
-			err = multierr.Append(err, commonconfig.ErrMissing{Name: "WSURL", Msg: "required for primary nodes"})
-		}
-	} else if n.WSURL.IsZero() {
-		if !sendOnly {
-			err = multierr.Append(err, commonconfig.ErrEmpty{Name: "WSURL", Msg: "required for primary nodes"})
-		}
-	} else {
+	// WSURL can be optional when LogBroadcaster is disabled, if WSURL not provided, rpc will return error for SubscribeFilterLogs
```
Comment (jmank88) [inline core/chains/evm/config/toml/config.go]: This validation needs to remain in place when `LogBroadcaster` is enabled. We make an effort to fail-fast and block node startup for invalid configuration, and relaxing that would be a regression.

**Review by jmank88 — CHANGES_REQUESTED** [review]

File: core/chains/evm/client/evm_client.go [hunk 2]
```diff
@@ -20,7 +20,7 @@ func NewEvmClient(cfg evmconfig.NodePool, chainCfg commonclient.ChainConfig, cli
 	var sendonlys []commonclient.SendOnlyNode[*big.Int, RPCClient]
 	largePayloadRPCTimeout, defaultRPCTimeout := getRPCTimeouts(chainType)
 	for i, node := range nodes {
-		if node.SendOnly != nil && *node.SendOnly {
+		if node.WSURL == nil || (node.SendOnly != nil && *node.SendOnly) {
```
Comment (dhaidashenko) [inline core/chains/evm/client/evm_client.go]: A secondary or SendOnly node has limited functionality. It's only used to broadcast transactions (fire and forget), and we are not running health checks for it.
We should not change the type of the node. Instead, introduce changes to the RPC client.

**Review by dhaidashenko — CHANGES_REQUESTED** [review]

File: core/chains/evm/client/config_builder.go [hunk 3]
```diff
@@ -76,18 +85,37 @@ func NewClientConfigs(
 	return chainConfig, nodePoolCfg, nodes, nil
 }
 
+// verifyLogBroadcasterFlag checks node config and return error if LogBroadcaster enabled but some node missing WSURL
+func verifyLogBroadcasterFlag(nodes []*toml.Node, LogBroadcasterEnabled bool) error {
```
Comment (dhaidashenko) [inline core/chains/evm/client/config_builder.go]: Config Builder is built for external applications like Atlas, where LogBroadcaster does not exist. The CL Node does not use it. 
`verifyLogBroadcasterFlag` should be moved to CL Node config validation, which is located in [toml pkg](https://github.com/smartcontractkit/chainlink/blob/1ea9f79793f646977b44e38a34b2e70c28b2849e/core/chains/evm/config/toml/config.go#L291)

Comment (huangzhen1997) [reply core/chains/evm/client/config_builder.go]: Do we need to check the `LogBrodcasterEnabled` flag for node during [validateConfig()](https://github.com/smartcontractkit/chainlink/blob/1ea9f79793f646977b44e38a34b2e70c28b2849e/core/chains/evm/config/toml/config.go#L971) ?

Comment (huangzhen1997) [reply core/chains/evm/client/config_builder.go]: Okay I moved the restriction (ws exist when `logBroadcaster` disabled + http exist when ws missing) to `rpc_client.go`.

Just making sure, we don't have to worry about the node config validation [here](https://github.com/smartcontractkit/chainlink/blob/1ea9f79793f646977b44e38a34b2e70c28b2849e/core/chains/evm/config/toml/config.go#L971) (also [here](https://github.com/smartcontractkit/chainlink/blob/1ea9f79793f646977b44e38a34b2e70c28b2849e/core/chains/evm/config/toml/config.go#L314)), because it's for external users right ?

File: core/chains/evm/client/evm_client.go (hunk 2 above)
Comment (huangzhen1997) [reply core/chains/evm/client/evm_client.go]: Moving the enforcement check to `Dial()` at rpc_client.go

File: core/chains/evm/client/rpc_client.go [hunk 4]
```diff
@@ -174,6 +179,7 @@ func NewRPCClient(
 	r.tier = tier
 	r.ws.uri = wsuri
 	r.finalizedBlockPollInterval = finalizedBlockPollInterval
+	r.logBroadcasterEnabled = logBroadcasterEnabled
 	if httpuri != nil {
```
Comment (dhaidashenko) [inline core/chains/evm/client/rpc_client.go:183]: RPC Client does not need to know about LogBroadcaster, this breaks abstraction

File: core/chains/evm/client/config_builder.go (hunk 3 above)
Comment (dhaidashenko) [reply core/chains/evm/client/config_builder.go]: Not sure why you've moved the config validation to the `rpc_client.go`, the link I've provided in my previous msg points to `core/chains/evm/config/toml/config.go`. 
The validation must be part of the config validation step to ensure CL Node won't start with the invalid configs

**Review by dhaidashenko — CHANGES_REQUESTED** [review]

File: core/config/docs/chains-evm.toml [hunk 5]
```diff
@@ -464,7 +464,7 @@ ObservationGracePeriod = '1s' # Default
 [[EVM.Nodes]]
 # Name is a unique (per-chain) identifier for this node.
 Name = 'foo' # Example
-# WSURL is the WS(S) endpoint for this node. Required for primary nodes.
+# WSURL is the WS(S) endpoint for this node. Can be optional when `LogBroadcasterEnabled` is `false`
```
Comment (jmank88) [inline core/config/docs/chains-evm.toml]: How about this format:
```suggestion
# WSURL is the WS(S) endpoint for this node. Required for primary nodes when `LogBroadcasterEnabled` is `true`
```

File: .changeset/silly-lies-boil.md [hunk 6]
```diff
@@ -0,0 +1,8 @@
+---
+"chainlink": minor
+---
+
+Make websocket URL flag `WSURL` for `EVM.Nodes`, and apply logic so that:
+* If WS URL was not provided, SubscribeFilterLogs should fail with an explicit error
+* If WS URL was not provided LogBroadcaster should be disabled 
+#internal
```
Comment (jmank88) [inline .changeset/silly-lies-boil.md]: Isn't this important for node ops?
```suggestion
#nops
```

File: core/chains/evm/client/config_builder.go [hunk 7]
```diff
@@ -79,15 +79,21 @@ func NewClientConfigs(
...
+		var wsURL, httpURL *commonconfig.URL
+		// wsUrl is optional, if LogBroadcaster is enabled, at least one wsURL is required and this will be checked in EVMConfig validation
+		if nodeCfg.WSURL != nil {
+			wsURL = commonconfig.MustParseURL(*nodeCfg.WSURL)
 		}
-		wsUrl := commonconfig.MustParseURL(*nodeCfg.WSURL)
-		httpUrl := commonconfig.MustParseURL(*nodeCfg.HTTPURL)
+
+		if nodeCfg.HTTPURL == nil {
+			return nil, fmt.Errorf("node config [%d]: HTTP URL", i)
```
Comment (jmank88) [inline core/chains/evm/client/config_builder.go]:
```suggestion
			return nil, fmt.Errorf("node config [%d]: missing HTTP URL", i)
```

File: core/chains/evm/client/rpc_client_test.go [hunk 8]
```diff
@@ -209,6 +209,16 @@ func TestRPCClient_SubscribeFilterLogs(t *testing.T) {
 	lggr := logger.Test(t)
 	ctx, cancel := context.WithTimeout(tests.Context(t), tests.WaitTimeout(t))
 	defer cancel()
+	t.Run("Failed SubscribeFilterLogs when WSURL is empty", func(t *testing.T) {
+		// ws is optional when LogBroadcaster is disabled, however SubscribeFilterLogs will return error if ws is missing
+		httpURL := url.URL{}
+		observedLggr, _ := logger.TestObserved(t, zap.DebugLevel)
+		rpcClient := client.NewRPCClient(observedLggr, url.URL{}, &httpURL, "rpc", 1, chainId, commonclient.Primary, 0, commonclient.QueryTimeout, commonclient.QueryTimeout, "")
```
Comment (jmank88) [inline core/chains/evm/client/rpc_client_test.go]: Could `httpURL` be omitted?
```suggestion
		observedLggr, _ := logger.TestObserved(t, zap.DebugLevel)
		rpcClient := client.NewRPCClient(observedLggr, url.URL{}, nil, "rpc", 1, chainId, commonclient.Primary, 0, commonclient.QueryTimeout, commonclient.QueryTimeout, "")
```

File: core/chains/evm/config/toml/config.go [hunk 9]
```diff
@@ -310,6 +313,12 @@ func (c *EVMConfig) ValidateConfig() (err error) {
 	if len(c.Nodes) == 0 {
 		err = multierr.Append(err, commonconfig.ErrMissing{Name: "Nodes", Msg: "must have at least one node"})
 	} else {
+		if c.LogBroadcasterEnabled != nil {
+			if err = verifyLogBroadcasterFlag(c.Nodes, *c.LogBroadcasterEnabled); err != nil {
+				err = multierr.Append(err, commonconfig.ErrMissing{Name: "Nodes", Msg: "must have at least one ws uri when LogBroadcaster is enabled"})
```
Comment (jmank88) [inline core/chains/evm/config/toml/config.go]:
```suggestion
				err = multierr.Append(err, commonconfig.ErrMissing{Name: "Nodes", Msg: "must have at least one node with a WSURL when LogBroadcaster is enabled"})
```

File: core/services/chainlink/config_test.go [hunk 10]
```diff
@@ -1331,7 +1331,9 @@ func TestConfig_Validate(t *testing.T) {
 		- 1.ChainID: invalid value (1): duplicate - must be unique
 		- 0.Nodes.1.Name: invalid value (foo): duplicate - must be unique
 		- 3.Nodes.4.WSURL: invalid value (ws://dupe.com): duplicate - must be unique
-		- 0: 3 errors:
+		- 0: 5 errors:
+			- logbroadcaster cannot be enabled unless all nodes have WSURL provided
```
Comment (jmank88) [inline core/services/chainlink/config_test.go:1335]: Is this true that _all_ nodes must have WSURL? The error message in toml/config.go says: `must have at least one`

File: core/services/chainlink/config_test.go [hunk 11]
```diff
@@ -1331,7 +1331,9 @@ func TestConfig_Validate(t *testing.T) {
...
+			- Nodes: missing: must have at least one ws uri when LogBroadcaster is enabled
```
Comment (jmank88) [inline core/services/chainlink/config_test.go]: These seem contradictory. Must all nodes have WSURL, or just one?

File: core/services/chainlink/testdata/config-invalid.toml [hunk 12]
```diff
@@ -83,6 +83,7 @@ MinIncomingConfirmations = 0
 
 [[EVM]]
 ChainID = '99'
+LogBroadcasterEnabled = false
```
Comment (jmank88) [inline core/services/chainlink/testdata/config-invalid.toml]: Shouldn't this be true in order to trigger the validation and produce errors? Setting false makes it more permissive of the node URL values.

File: core/chains/evm/client/rpc_client_test.go (hunk 8 above)
Comment (huangzhen1997) [reply core/chains/evm/client/rpc_client_test.go]: For this specific test it can't, since Dial() will fail due to `errWSAndHTTPBothMissing`. I can add another test that covers when both WS and HTTPURL are empty

File: core/services/chainlink/config_test.go (hunk 11 above)
Comment (huangzhen1997) [reply core/services/chainlink/config_test.go]: Thanks for catching it. According to the requirement, I was intended to enforce `Must all nodes have WSURL when LogBroadcaster is enabled`, so the second error message is incorrect. But I will double check if the requirement is correct, as it might not backward compatible

File: core/services/chainlink/testdata/config-invalid.toml (hunk 12 above)
Comment (huangzhen1997) [reply core/services/chainlink/testdata/config-invalid.toml]: For this specific chain setting I set it to `false`, so the result will not include the WS URL missing error. Otherwise modification to config_test.go at line 1367 will be expected.

Comment (jmank88) [reply core/services/chainlink/testdata/config-invalid.toml]: This file is specifically meant to provoke errors via invalid configuration. We should flip this to true in order to have coverage on the additional validation.

Comment (huangzhen1997) [reply core/services/chainlink/testdata/config-invalid.toml]: Okay sounds good. Removed the false assignment, as default value is `true`

File: core/services/chainlink/config_test.go (hunk 11 above)
Comment (huangzhen1997) [reply core/services/chainlink/config_test.go]: Updated the error message. We going to require all primary nodes to provide valid WS URL when LogBroadcaster is enabled.

File: .changeset/silly-lies-boil.md [hunk 13]
```diff
@@ -0,0 +1,8 @@
+---
+"chainlink": minor
+---
+
+Make websocket URL flag `WSURL` for `EVM.Nodes`, and apply logic so that:
```
Comment (dhaidashenko) [inline .changeset/silly-lies-boil.md]:
```suggestion
Make websocket URL `WSURL` for `EVM.Nodes` optional, and apply logic so that:
```

File: core/chains/evm/client/config_builder.go [hunk 14]
```diff
@@ -79,15 +79,21 @@ func NewClientConfigs(
 func parseNodeConfigs(nodeCfgs []NodeConfig) ([]*toml.Node, error) {
 	nodes := make([]*toml.Node, len(nodeCfgs))
 	for i, nodeCfg := range nodeCfgs {
-		if nodeCfg.WSURL == nil || nodeCfg.HTTPURL == nil {
-			return nil, fmt.Errorf("node config [%d]: missing WS or HTTP URL", i)
+		var wsURL, httpURL *commonconfig.URL
+		// wsUrl is optional, if LogBroadcaster is enabled, at least one wsURL is required and this will be checked in EVMConfig validation
```
Comment (dhaidashenko) [inline core/chains/evm/client/config_builder.go]: `LogBroadcaster` is not relevant to the context of `evm/client`. It's one of many users, we should not reference it here.

File: core/chains/evm/client/rpc_client.go [hunk 15]
```diff
@@ -195,30 +197,35 @@ func (r *rpcClient) Dial(callerCtx context.Context) error {
 	ctx, cancel := r.makeQueryCtx(callerCtx, r.rpcTimeout)
 	defer cancel()
 
-	promEVMPoolRPCNodeDials.WithLabelValues(r.chainID.String(), r.name).Inc()
-	lggr := r.rpcLog.With("wsuri", r.ws.uri.Redacted())
-	if r.http != nil {
-		lggr = lggr.With("httpuri", r.http.uri.Redacted())
+	if r.ws.uri.String() == "" && r.http == nil {
```
Comment (dhaidashenko) [inline core/chains/evm/client/rpc_client.go]: WS is optional. Can't we make it a `*url.URL` instead of `url.URL`?

File: core/chains/evm/client/rpc_client.go [hunk 16]
```diff
@@ -195,30 +197,35 @@ func (r *rpcClient) Dial(callerCtx context.Context) error {
...
+			promEVMPoolRPCNodeDialsFailed.WithLabelValues(r.chainID.String(), r.name).Inc()
+			return r.wrapRPCClientError(pkgerrors.Wrapf(err, "error while dialing websocket: %v", r.ws.uri.Redacted()))
+		}
 
-	r.ws.rpc = wsrpc
-	r.ws.geth = ethclient.NewClient(wsrpc)
+		r.ws.rpc = wsrpc
+		r.ws.geth = ethclient.NewClient(wsrpc)
+	} else {
+		lggr = lggr.With("wsuri", "empty WS URI")
```
Comment (dhaidashenko) [inline core/chains/evm/client/rpc_client.go]: There is no need to include `empty was URI` into the log

File: core/chains/evm/client/rpc_client_test.go [hunk 17]
```diff
@@ -45,6 +45,7 @@ func TestRPCClient_SubscribeNewHead(t *testing.T) {
 
 	chainId := big.NewInt(123456)
 	lggr := logger.Test(t)
+	var missingBothWSAndHTTPErr = errors.New("cannot dial rpc client when both ws and http info are missing")
```
Comment (dhaidashenko) [inline core/chains/evm/client/rpc_client_test.go]: We should try minimizing the lifespan of a variable. This one is only used in a single test case. There is no need to define it globally for the testing function.

File: core/chains/evm/client/rpc_client_test.go [hunk 18]
```diff
@@ -209,6 +216,15 @@ func TestRPCClient_SubscribeFilterLogs(t *testing.T) {
...
 	ctx, cancel := context.WithTimeout(tests.Context(t), tests.WaitTimeout(t))
 	defer cancel()
+	t.Run("Failed SubscribeFilterLogs when WSURL is empty", func(t *testing.T) {
+		// ws is optional when LogBroadcaster is disabled, however SubscribeFilterLogs will return error if ws is missing
+		observedLggr, _ := logger.TestObserved(t, zap.DebugLevel)
+		rpcClient := client.NewRPCClient(observedLggr, url.URL{}, &url.URL{}, "rpc", 1, chainId, commonclient.Primary, 0, commonclient.QueryTimeout, commonclient.QueryTimeout, "")
+		require.Nil(t, rpcClient.Dial(ctx))
+
+		_, err := rpcClient.SubscribeFilterLogs(ctx, ethereum.FilterQuery{}, make(chan types.Log))
+		require.Error(t, err)
```
Comment (dhaidashenko) [inline core/chains/evm/client/rpc_client_test.go]: Could you ensure that we return proper error msg?

File: core/chains/evm/client/rpc_client.go [hunk 19]
```diff
@@ -75,6 +75,8 @@ var (
 			float64(8 * time.Second),
 		},
 	}, []string{"evmChainID", "nodeName", "rpcHost", "isSendOnly", "success", "rpcCallName"})
+	errSubscribeFilterLogsNotAllowedWithoutWS = errors.New("SubscribeFilterLogs is not allowed without ws url")
```
Comment (dhaidashenko) [inline core/chains/evm/client/rpc_client.go]: Are we reusing these variables across multiple functions? If not, it might make sense to define them inline or make them public and use in tests

File: core/chains/evm/config/toml/config.go [hunk 20]
```diff
@@ -311,13 +314,26 @@ func (c *EVMConfig) ValidateConfig() (err error) {
...
+		}
+
 		for _, n := range c.Nodes {
 			if n.SendOnly != nil && *n.SendOnly {
 				continue
 			}
-			hasPrimary = true
-			break
+
+			if !hasPrimary {
```
Comment (dhaidashenko) [inline core/chains/evm/config/toml/config.go]: nit: if statement is redundant as it's fine to set `hasPrimary = true` if it's already true

File: core/chains/evm/config/toml/config.go [hunk 21]
```diff
@@ -311,13 +314,26 @@ func (c *EVMConfig) ValidateConfig() (err error) {
...
+				hasPrimary = true
+			}
+
+			// if the node is a primary node, then the WS URL is required when LogBroadcaster is enabled
+			if logBroadcasterEnabled && (n.WSURL == nil || n.WSURL.IsZero()) {
+				err = multierr.Append(err, commonconfig.ErrMissing{Name: "Nodes", Msg: "all primary nodes must provide a valid WSURL when LogBroadcaster is enabled"})
```
Comment (dhaidashenko) [inline core/chains/evm/config/toml/config.go]: It might be a good idea to specify failed node index

**Review by dhaidashenko — CHANGES_REQUESTED** [review]
Could you try starting the CL node locally without WS URL? It seems like in the current version, it should fail with nil pointer deference.  Due to [this](https://github.com/smartcontractkit/chainlink/blob/7738ff0195a1d8447324d3577b3c6c24e65b1dbf/core/chains/evm/client/evm_client.go#L30)

File: core/chains/evm/client/rpc_client.go (hunk 15 above)
Comment (huangzhen1997) [reply core/chains/evm/client/rpc_client.go]: We can, but it requires a lot refactor and adding nil ptr check and not sure if we want to do in this PR

Comment (dhaidashenko) [reply core/chains/evm/client/rpc_client.go]: Hm, I'm not sure that we'll have a better chance to do the change. 
Just to clarify, we need to change wsuri's type to *url.URL in the NewRPCClient method. As at the moment it's not clear from the method signature that it's optional

File: core/chains/evm/config/toml/config.go [hunk 22]
```diff
@@ -311,13 +313,24 @@ func (c *EVMConfig) ValidateConfig() (err error) {
...
+
+			// if the node is a primary node, then the WS URL is required when LogBroadcaster is enabled
+			if logBroadcasterEnabled && (n.WSURL == nil || n.WSURL.IsZero()) {
+				err = multierr.Append(err, commonconfig.ErrMissing{Name: "Nodes", Msg: fmt.Sprintf("%vth node is a primary node and it's missing valid WSURL with LogBroadcaster is enabled", i)})
+			}
 		}
+
 		if !hasPrimary {
 			err = multierr.Append(err, commonconfig.ErrMissing{Name: "Nodes",
 				Msg: "must have at least one primary node with WSURL"})
```
Comment (dhaidashenko) [inline core/chains/evm/config/toml/config.go]:
```suggestion
				Msg: "must have at least one primary node"})
```

File: core/chains/evm/config/toml/config.go [hunk 23]
```diff
@@ -311,13 +313,24 @@ func (c *EVMConfig) ValidateConfig() (err error) {
...
 			if n.SendOnly != nil && *n.SendOnly {
 				continue
 			}
+
 			hasPrimary = true
-			break
+
+			// if the node is a primary node, then the WS URL is required when LogBroadcaster is enabled
+			if logBroadcasterEnabled && (n.WSURL == nil || n.WSURL.IsZero()) {
+				err = multierr.Append(err, commonconfig.ErrMissing{Name: "Nodes", Msg: fmt.Sprintf("%vth node is a primary node and it's missing valid WSURL with LogBroadcaster is enabled", i)})
```
Comment (dhaidashenko) [inline core/chains/evm/config/toml/config.go]:
```suggestion
				err = multierr.Append(err, commonconfig.ErrMissing{Name: "Nodes", Msg: fmt.Sprintf("%vth primary node must have a valid WSURL when LogBroadcaster is enabled", i)})
```

File: core/chains/evm/client/evm_client.go [hunk 24]
```diff
@@ -15,22 +15,25 @@ import (
...
 	var sendonlys []commonclient.SendOnlyNode[*big.Int, RPCClient]
 	largePayloadRPCTimeout, defaultRPCTimeout := getRPCTimeouts(chainType)
 	for i, node := range nodes {
+		var ws *url.URL
+		if node.WSURL != nil {
+			ws = (*url.URL)(node.WSURL)
+		}
 		if node.SendOnly != nil && *node.SendOnly {
-			rpc := NewRPCClient(lggr, empty, (*url.URL)(node.HTTPURL), *node.Name, int32(i), chainID,
+			rpc := NewRPCClient(lggr, ws, (*url.URL)(node.HTTPURL), *node.Name, int32(i), chainID,
```
Comment (dhaidashenko) [inline core/chains/evm/client/evm_client.go]:
```suggestion
			rpc := NewRPCClient(lggr, nil, (*url.URL)(node.HTTPURL), *node.Name, int32(i), chainID,
```

Previously, we always passed empty URLs for sendOnly RPC clients. To avoid unexpected behavior, we should pass nil instead of the URL specified in the config. 
In theory, NOPs might have specified unreachable URL/empty URL, which we've previously ignored.
But with this change and changes from Dylans PR will start using it, which could cause issue.

**Review by dhaidashenko — CHANGES_REQUESTED** [review]

File: core/chains/evm/config/toml/config.go (hunk 23 above)
Comment (huangzhen1997) [reply core/chains/evm/config/toml/config.go]: `%vth node` instead of `%vth primary` node is more accurate. Or `%vth node (primary)`

**Review by jmank88 — DISMISSED** [review]

**Review by dhaidashenko — APPROVED** [review]

**Review by jmank88 — APPROVED** [review]

**Review by DylanTinianov — APPROVED** [review]
